src/map.rs

```diff
@@ -63,72 +74,79 @@ pub type MapResult<T> = Result<T, MapError>;
-    ///
-    /// # Warning
-    /// May never be used, and may be removed.
-    Refresh {
-        /// The `Handle` of the `Chunk`.
+        /// The Handle of the Chunk.
         handle: Handle<Chunk>,
     },
-    /// If the chunk had been modified.
-    Modified {
+    /// An event when a layer is created for all chunks.
+    AddedLayer {
+        z: usize,
+        kind: LayerKind,
+    },
+    MovedLayer {
+        from_z: usize,
+        to_z: usize,
+    },
+    /// An event when a layer is removed for all chunks.
+    RemovedLayer {
+        z: usize,
+    },
+    /// An event when a chunk had been modified by changing tiles.
+    ModifiedChunk {
         /// The map index where the chunk needs to be stored.
-        index: usize,
+        handle: Handle<Chunk>,
         /// The `TileSetter` that is used to set all the tiles.
         setter: TileSetter,
     },
-    /// If the chunk needs to be despawned.
-    Despawned {
+    /// An event when a chunk is spawned.
+    SpawnedChunk {
+        handle: Handle<Chunk>,
+    },
+    /// If the chunk needs to be despawned, this event is used.
+    DespawnedChunk {
         /// The `Handle` of the `Chunk`.
         handle: Handle<Chunk>,
-        /// The `Entity` that needs to be despawned.
-        entity: Entity,
+        // /// The `Entity` that needs to be despawned.
+        // entity: Entity,
     },
     /// If the chunk needs to be removed.
     ///
     /// # Warning
     /// This is destructive action! All data will be dropped and removed.
-    Removed {
+    RemovedChunk {
         /// The map index where the chunk needs to be removed.
-        index: usize,
-        /// The `Entity` that needs to be despawned.
-        entity: Entity,
+        handle: Handle<Chunk>,
+        // /// The `Entity` that needs to be despawned.
+        // entity: Entity,
     },
 }
 
 /// A TileMap which maintains chunks and its tiles within.
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
-#[derive(Debug, RenderResources)]
+#[derive(Debug)]
 pub struct TileMap {
-    #[render_resources(ignore)]
     dimensions: Vec2,
     chunk_dimensions: Vec3,
-    #[render_resources(ignore)]
     tile_dimensions: Vec2,
+    current_depth: usize,
+    layers: Vec<Option<LayerKind>>,
     #[cfg_attr(feature = "serde", serde(skip))]
-    // Should change to HashSet when merged into bevy
-    #[render_resources(ignore)]
-    chunks: Vec<Option<Handle<Chunk>>>,
+    chunks: HashMap<usize, Handle<Chunk>>,
     #[cfg_attr(feature = "serde", serde(skip))]
-    #[render_resources(ignore)]
-    entities: HashMap<usize, Entity>,
+    entities: HashMap<usize, Vec<Entity>>,
     #[cfg_attr(feature = "serde", serde(skip))]
-    #[render_resources(ignore)]
     events: Events<MapEvent>,
     #[cfg_attr(feature = "serde", serde(skip))]
-    #[render_resources(ignore)]
     texture_atlas: Handle<TextureAtlas>,
+    // NOTE: If there is a better way to keep track of spawned chunks, that
```

Comment:
I don't see why you have this field at all? You add things to it and remove them later but you never actually use the data for anything.

Reply:
There is supposed to be public API to get that data.

Reply:
Can you just rely on Bevy's `Children`? Since the chunks will all be children of the TileMap.